Feature/async queue

Aprogiena opened the pull request:
Async queue is a simple thread-safe queue that asynchronously executes a user-defined callback when a new item is added to the queue.

Implementation of async queue that we want for the node job part 19, which will fix issue #991 by refactoring that area of the code. 

Tests are included.

There are no changes to functionality of any existing components in this PR.

dangershony commented:
I thought @noescape will work on this.
I like the name but it's not so correct its an AsyncTrigger more then a queue.

Aprogiena commented:
at the heart of it, it is a queue. it is a queue because the producer enqueues items in specific order and the consumer consumes them in that order.

currently a similar mechanism is used in NetworkPeer code and there it is called EventLoopMessageListener, in which a message producer adds messages to listener queues

so the primary purpose of AsyncQueue is to replace EventLoopMessageListener, which uses one thread per listener

the optional feature that allows you to manually trigger the consumer is there in order to allow us to create a second class that could be built on the top of this one with inheritance, that will trigger the manual signal with a timer. that can be then used for the blockstore thing, which is what @noescape00 was working on (and is still kind of working on)

Aprogiena commented:
so basically i need this for the node job part 19
and it is possible that it will be a multi-purpose class (i.e. reused in other components as well)

@noescape00 won't be able to deliver a similar class quickly enough for me, so although there is no rush with blockstore refactoring, i would like to advance in the node job much earlier

noescape00 commented:
>@noescape00 won't be able to deliver a similar class quickly enough for me

It's already done and I can wrap up a PR right now if you'd like. =)

noescape00 commented:
What is the purpose if this class? 
Basically it's a wrapper around a queue that has triggers on enqueue and a manual one. 
Why manual trigger is there? It's not connected to a queue in any way, it doesn't even have an access to the queue.
 
I don't think that having in this class an anonymous function that can be manually executed is any better than letting a user do this if he want's too. 
I just don't see a scenario where having this trigger would be actually better than doing it in user code.

noescape00 commented:
I've created a PR with classic implementation of async queue: https://github.com/stratisproject/StratisBitcoinFullNode/pull/1013

noescape00 commented:
@dangershony we've been arguing with @Aprogiena  on the design of this thing for a week so we need a second opinion on this =)

Basically both of the approaches are used to solve same problem- make implementing a scenario where timer and trigger are involved easier. 
A concrete usecase for this is the relay worker: we wanna form batches by waiting for a second or so between announcing the headers BUT if we get a tip header it means we need to send it now because it's critical to propagate a new block right away (especially if we've mined the block). 

My position was to make this easy in the user code even at a cost of a bit more difficult class. In the end I had a class which usage is easy and straight forward: 

```
AsyncConsumerQueue<int> processedQueue = new AsyncConsumerQueue<int>(
    innerQueue =>
    {
        // Process the queue. This is the only place with the access to the 'innerQueue'
    }, item =>
    {
        // Condition that should be met in order to start processing the
        // queue right after an item as added. 
        return item == *;
    }, TimeSpan.FromMilliseconds(500)); // Timer delay. Queue processing will be invoked in 0.5 sec if the condition was not met. 
```

@Aprogiena's idea was to make the queue class as tiny as possible even at a cost of making user code more difficult and forcing user to understand queue implementation before using it.

Aprogiena commented:
I'm OK to remove the manual triggering from the async queue as it is not needed for me node job refactoring. It is there just as a concept to be used in the future that we can built blockstore thing on. If we don't want to do that, it is fine, I can remove the manual trigger.

Aprogiena commented:
@noescape00 the purpose is described above - to replace EventLoopMessageListener which is blocking queue consuming a thread.

Aprogiena commented:
@dangershony so i've removed the controversial manual trigger, so you should not have problem with the name of the class now :-)

@noescape00 i've removed the controversial manual trigger, so you should not have a feeling that this is supposed to be used in blockstore and compete with your implementation

Aprogiena commented:
For the sake of discussion on extension towards the timer trigger, i've left the original version with manual trigger in https://github.com/Aprogiena/StratisBitcoinFullNode/tree/feature/async-queue-with-manual-trigger if anyone was interested

noescape00 commented:
>the purpose is described above - to replace EventLoopMessageListener which is blocking queue consuming a thread.

Which can also be done with AsyncConsumerQueue in more simple and clean way like this: 

```
AsyncConsumerQueue<int> processedQueue = new AsyncConsumerQueue<int>(
    innerQueue =>
    {
        NetworkPeerServer.ProcessMessageAsync()
    });
```

This will also avoid a thread consumption. So basically there is no need for another queue that can only one use case if we can use a more general class that can support exactly the behavior you require.

noescape00 commented:
>@noescape00 i've removed the controversial manual trigger, so you should not have a feeling that this is supposed to be used in blockstore and compete with your implementation

It still says in the description that it IS supposed to. 

Actually even with this in mind I don't feel like we need this. Why bringing more narrow-usage classes to the codebase when we can have just 1 class that can be well used in many usecases including the one that made you create this PR?

noescape00 commented on src/Stratis.Bitcoin/Utilities/AsyncQueue.cs:
```diff
@@ -0,0 +1,104 @@
+        /// Initializes the queue.
+        /// </summary>
+        /// <param name="onEnqueueAsync">Callback routine to be called when a new item is added to the queue.</param>
+        public AsyncQueue(OnEnqueueAsync onEnqueueAsync)
```
Why not using `Func<T, CancellationToken, Task>` instead of creating a delegate?

Aprogiena replied:
the ONLY reason for that is to have comments on the arguments of the delegate, 
I originally did have Func there and found out i lack the documentation of those parameters

i checked with MSDN and they do use delegates for the same reason (at least i believe), those delegates are commented separately and that's what i needed here

noescape00 commented on src/Stratis.Bitcoin/Utilities/AsyncQueue.cs:
```diff
@@ -0,0 +1,104 @@
+        public delegate Task OnEnqueueAsync(T item, CancellationToken cancellationToken);
+
+        /// <summary>Storage of items in the queue that are waiting to be consumed.</summary>
+        private readonly ConcurrentQueue<T> items;
```
Can you please make it `protected`?

Aprogiena replied:
if needed, the further work can change this

just as there is no need for manual trigger now, it is perfectly fine for this to be private

there is no problem changing anything in the class later and extend it to either that version with manual trigger that i have in separate branch, or to whatever you have in your mind. this being private does not forbid any changes later

noescape00 replied:
Ok, reasonable.

noescape00 commented:
We've came to an agreement that I'll build queue with timer&trigger on top of the AsyncQueue and this one should be pushed because it's required for node refactoring and is time sensitive.

dangershony commented:
This looks good to me.
